### DISCUSSION TITLE
Using Bytes to store binary

### DISCUSSION THREAD

**Discussion by cheetosysst** [prisma/prisma#13301] [Q&A] [answered]
I'm using next.js, mongodb and prisma to make a simple demo for school project.
I want to store binary files in db, so I wrote the following
```prisma
model Book {
  id String @id @default(auto()) @map("_id") @db.ObjectId
  title String
  title_alternate String
  authors String[]
  count Int
  description String
  eprice Int
  price Int
  tag String[]
  markdown String
  image Bytes
}
```
```js
const getImageFile = () => {
	return new Promise((acc, err) => {
		const reader = new FileReader();

		reader.onload = (event) => {
			console.log((event.target.result));
			acc(event.target.result);
		};
		reader.onerror = (error) => {
			err(error);
		};

		reader.readAsArrayBuffer(event.target.image.files[0]);
	});
};
var imageFile = await getImageFile();
```
The results is an ArrayBuffer, but prisma complained
```
Argument image: Got invalid value {} on prisma.createOneBook. Provided Json, expected Bytes.
```
And I couldn't figure out what "Bytes" it's expecting. I know I can't pass `imageFile` directly because that's just the ArrayBuffer not the actual data? What am I missing?

**Comment by austincrim** [comment, marked as answer]
[`Bytes` fields expect a JavaScript `Buffer` type.](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-fields#working-with-bytes) If you have an `ArrayBuffer`, you need to convert it to a `Buffer` before passing it to your Prisma query.
```ts
let buffer = Buffer.from(imageFile)
```

**Reply by cheetosysst** [reply]
Hi, thanks for the reply.
I tried
```js
// Client side
var imageFile = await getImageFile();
var imageBuffer = Buffer.from(imageFile); 

let data = {
	image: imageBuffer,
};
```
and on the api side
```js
const {
	image,
} = req.body;

const createBook = await prisma.Book.create({
	data: {
		image: image,
	},
});
```
It still have the same 'expected bytes received json' error, I tried sending `imageFile` to the api and process it there but it ends up with "write after end error"
```js
const createBook = await prisma.Book.create({
	data: {
		image: Buffer.from(image),
	},
});
```

**Reply by austincrim** [reply]
How are you sending the `imageFile` from the client to the server? You can't serialize an `ArrayBuffer` to JSON so you are most likely losing it during the request. 

You can read more [here](https://stackoverflow.com/questions/48291288/how-to-send-array-buffer-data-along-with-string-json-to-nodejs-server) about how to send JavaScript Buffers over HTTP.

**Reply by cheetosysst** [reply]
I was able to finally upload an image, but I end up using other methods to save files😅.
Thanks for the reply and resources.

**Reply by skabdulM** [reply]
i want to upload images what did you do to achieve tha

**Reply by cheetosysst** [reply]
@skabdulM 
I end up uploading images to other places, and only save the name of the file to database. AFAIK it's really not a great idea to save images in a database.

**Reply by skabdulM** [reply]
I ended up creating another API to upload images and get their links

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
